source/OctoVersion.Core/VersionTemplates/VersionParser.cs

```diff
@@ -0,0 +1,203 @@
+            {
+                suffix += $".{GetValueForToken(octoVersionInfo, _tokenMap[i])}";
+            }
+
+            result = result.Substring(0, Math.Min(result.Length, 19 - suffix.Length)) + suffix;
+
+            return $"-{InvalidPreReleaseCharacters.Replace(result, "-")}";
+        }
+
+        string GetValueForToken(OctoVersionInfo octoVersionInfo, string token)
+        {
+            switch (token)
+            {
+                case "{major}":
+                    return octoVersionInfo.Major.ToString();
+                case "{minor}":
+                    return octoVersionInfo.Minor.ToString();
+                case "{patch}":
+                    return octoVersionInfo.Patch.ToString();
+                case "{preReleaseTag}":
+                    return octoVersionInfo.PreReleaseTag.ToString();
+                case "{build}":
+                    return octoVersionInfo.Build.ToString();
+                case "{metadata}":
+                    return octoVersionInfo.BuildMetadata.ToString();
+            }
+
+            return string.Empty;
+        }
+
+        static (string[] tokens, string metadata) SplitString(string template)
+        {
+            // if there is metadata it must always be at the end. It may contain . and -, so we split based in it first
+            var versionPlusMetadata = template.Split(new[] { "+" }, StringSplitOptions.None);
+
+            // the remainder of the version split is then done on what's before the metadata
+            var versionTokens = versionPlusMetadata
+                .First()
+                .Split(new[] { "." }, StringSplitOptions.None)
+                .SelectMany(x =>
+                {
+                    var preReleaseIndex = x.IndexOf("-", StringComparison.OrdinalIgnoreCase);
+                    if (preReleaseIndex < 0)
+                        return new[] { x };
+                    return new[] { x.Substring(0, preReleaseIndex), x.Substring(preReleaseIndex + 1) };
+                })
+                .ToList();
+
+            var metadata = string.Empty;
+            if (versionPlusMetadata.Length > 1)
+                metadata = versionPlusMetadata.Skip(1).First();
+
+            return (versionTokens.ToArray(), metadata);
+        }
+
+        int GetMarkerIndexInTemplate(string marker)
+        {
+            for (int i = 0; i < _tokenMap.Length; i++)
+                if (_tokenMap[i].Contains(marker))
+                    return i;
+            return -1;
+        }
+
+        string GetValue(string[] values, string marker)
+        {
+            var index = GetMarkerIndexInTemplate(marker);
+            if (index < 0 || index >= values.Length)
+                return string.Empty;
+            return values[index];
+        }
+
+        public SimpleVersion? TryParseSimpleVersion(string versionString)
+        {
+            try
+            {
+                if (versionString.StartsWith("v"))

```

Comment:
I don't think we should know about magic tags here. It's either semver or it's invalid.